File: tests/java/tests-java-gradle-junit5-allure3-selenium/src/test/java/pages/HomePage.java
package pages;

import api.AuthApiClient;
import helpers.Ui;
import io.qameta.allure.Step;
import org.openqa.selenium.WebElement;

public class HomePage extends BasePage<HomePage> {

    private static final String AUTH_TOKEN_KEY_JS =
            "var m=location.pathname.match(/\\/(backend-[^/]+)\\//);"
                    + "return m ? 'authToken:' + m[1] : 'authToken';";

    private static final String DELETE_ACCOUNT_CONFIRM =
            "Delete this account? This cannot be undone.";

    private String authTokenKey() {
        return String.valueOf(Ui.js(AUTH_TOKEN_KEY_JS));
    }

    private void stubConfirm(boolean accepted) {
        Ui.js(
                "window.__deleteConfirm = null;"
                        + "(function(accepted) {"
                        + "  window.confirm = function(msg) {"
                        + "    window.__deleteConfirm = msg;"
                        + "    return accepted;"
                        + "  };"
                        + "})(arguments[0]);",
                accepted);
    }

    private void shouldHaveConfirmMessage() {
        Ui.waitUntil(driver -> DELETE_ACCOUNT_CONFIRM.equals(Ui.js("return window.__deleteConfirm;")));
    }

    @Step("Open home page")
    public HomePage openPage() {
        Ui.open("/");
        return shouldBeOpen();
    }

    @Step("Open home page with local storage authentication")
    public HomePage openPageWithLocalStorageAuthentication(String username, String password) {
        String token = AuthApiClient.login(username, password);
        Ui.open("/login");
        Ui.js(
                "localStorage.setItem(arguments[0], arguments[1]);",
                authTokenKey(),
                token
        );
        Ui.open("/");
        return shouldBeOpen();
    }

    @Step("Open home page with invalid local storage token")
    public HomePage openPageWithInvalidToken() {
        Ui.open("/login");
        Ui.js(
                "localStorage.setItem(arguments[0], arguments[1]);",
                authTokenKey(),
                "invalid-token"
        );
        Ui.open("/");
        return shouldBeOpen();
    }

    @Override
    @Step("Verify home page is open")
    public HomePage shouldBeOpen() {
        Ui.shouldBeVisible("multistack-layout");
        return this;
    }

    @Step("Verify home layout is mounted")
    public HomePage shouldShowLayout() {
        Ui.shouldBeVisible("multistack-layout");
        Ui.shouldBeVisible("items-list");
        return this;
    }

    @Step("Verify home layout and health are mounted")
    public HomePage shouldShowLayoutAndHealth() {
        Ui.shouldBeVisible("multistack-layout");
        Ui.shouldBeVisible("health-status");
        return this;
    }

    @Step("Home layout panel is visible")
    public WebElement layoutPanel() {
        return Ui.el("multistack-layout");
    }

    @Step("Welcome panel is visible")
    public WebElement welcomePanelElement() {
        return Ui.el("welcome-panel");
    }

    @Step("Verify welcome panel stays hidden")
    public HomePage shouldHideWelcomePanel() {
        Ui.shouldHaveAttribute("welcome-panel", "hidden", "");
        return this;
    }

    @Step("Verify auth token was cleared from localStorage")
    public HomePage shouldClearAuthToken() {
        Ui.waitUntil(driver -> Ui.js("return localStorage.getItem(arguments[0]);", authTokenKey()) == null);
        return this;
    }

    @Step("Verify health status contains: {textFragment}")
    public HomePage shouldShowHealthText(String textFragment) {
        Ui.shouldHaveText("health-status", textFragment);
        return this;
    }

    @Step("Verify items list contains: {textFragment}")
    public HomePage shouldShowItemText(String textFragment) {
        Ui.shouldHaveText("items-list", textFragment);
        return this;
    }

    @Step("Verify items panel shows a readable error: {textFragment}")
    public HomePage shouldShowItemsError(String textFragment) {
        Ui.shouldHaveText("items-list", textFragment);
        return this;
    }

    @Step("Verify health panel shows a readable error: {textFragment}")
    public HomePage shouldShowHealthError(String textFragment) {
        Ui.shouldHaveText("health-status", textFragment);
        return this;
    }

    @Step("Verify welcome message: {message}")
    public HomePage shouldHaveWelcomeMessage(String message) {
        Ui.shouldBeVisible("welcome-panel");
        Ui.shouldHaveText("welcome-message", message);
        return this;
    }

    @Step("Verify session panel offers logout and delete account")
    public HomePage shouldShowSessionActions() {
        Ui.shouldBeVisible("logout-button");
        Ui.shouldHaveText("logout-button", "Logout");
        Ui.shouldBeVisible("delete-account-button");
        Ui.shouldHaveText("delete-account-button", "Delete account");
        return this;
    }

    @Step("Click logout button")
    public LoginPage clickLogoutButton() {
        Ui.click("logout-button");
        return new LoginPage();
    }

    @Step("Click delete account and confirm")
    public LoginPage clickDeleteAccountAndConfirm() {
        stubConfirm(true);
        Ui.click("delete-account-button");
        shouldHaveConfirmMessage();
        return new LoginPage();
    }

    @Step("Click delete account and cancel the confirm")
    public HomePage clickDeleteAccountAndCancel() {
        stubConfirm(false);
        Ui.click("delete-account-button");
        shouldHaveConfirmMessage();
        return this;
    }

    @Step("Verify auth token remains in localStorage")
    public HomePage shouldKeepAuthToken() {
        Ui.waitUntil(driver -> Ui.js("return localStorage.getItem(arguments[0]);", authTokenKey()) != null);
        return this;
    }
}

File: tests/csharp/tests-csharp-nunit-restsharp-selenium/Pages/HomePage.cs
using Allure.NUnit.Attributes;
using Api;
using Helpers;
using OpenQA.Selenium;

namespace Pages;

public sealed class HomePage : BasePage<HomePage>
{
    private const string AuthTokenKeyJs =
        "var m=location.pathname.match(/\\/(backend-[^/]+)\\//);"
        + "return m ? 'authToken:' + m[1] : 'authToken';";

    private const string DeleteAccountConfirm =
        "Delete this account? This cannot be undone.";

    private static string AuthTokenKey() => Convert.ToString(Ui.Js(AuthTokenKeyJs)) ?? "authToken";

    [AllureStep("Open home page")]
    public HomePage OpenPage()
    {
        Ui.Open("/");
        return ShouldBeOpen();
    }

    [AllureStep("Open home page with local storage authentication")]
    public HomePage OpenPageWithLocalStorageAuthentication(string username, string password)
    {
        var token = AuthApiClient.Login(username, password);
        Ui.Open("/login");
        Ui.Js("localStorage.setItem(arguments[0], arguments[1]);", AuthTokenKey(), token);
        Ui.Open("/");
        return ShouldBeOpen();
    }

    [AllureStep("Open home page with invalid local storage token")]
    public HomePage OpenPageWithInvalidToken()
    {
        Ui.Open("/login");
        Ui.Js("localStorage.setItem(arguments[0], arguments[1]);", AuthTokenKey(), "invalid-token");
        Ui.Open("/");
        return ShouldBeOpen();
    }

    [AllureStep("Verify home page is open")]
    public override HomePage ShouldBeOpen()
    {
        Ui.ShouldBeVisible("multistack-layout");
        return this;
    }

    [AllureStep("Verify home layout is mounted")]
    public HomePage ShouldShowLayout()
    {
        Ui.ShouldBeVisible("multistack-layout");
        Ui.ShouldBeVisible("items-list");
        return this;
    }

    [AllureStep("Verify home layout and health are mounted")]
    public HomePage ShouldShowLayoutAndHealth()
    {
        Ui.ShouldBeVisible("multistack-layout");
        Ui.ShouldBeVisible("health-status");
        return this;
    }

    [AllureStep("Home layout panel is visible")]
    public IWebElement LayoutPanel() => Ui.El("multistack-layout");

    [AllureStep("Welcome panel is visible")]
    public IWebElement WelcomePanelElement() => Ui.El("welcome-panel");

    [AllureStep("Verify welcome panel stays hidden")]
    public HomePage ShouldHideWelcomePanel()
    {
        Ui.ShouldHaveAttribute("welcome-panel", "hidden", "");
        return this;
    }

    [AllureStep("Verify auth token was cleared from localStorage")]
    public HomePage ShouldClearAuthToken()
    {
        Ui.WaitUntil(_ => Ui.Js("return localStorage.getItem(arguments[0]);", AuthTokenKey()) == null ? true : (bool?)null);
        return this;
    }

    [AllureStep("Verify health status contains: {textFragment}")]
    public HomePage ShouldShowHealthText(string textFragment)
    {
        Ui.ShouldHaveText("health-status", textFragment);
        return this;
    }

    [AllureStep("Verify items list contains: {textFragment}")]
    public HomePage ShouldShowItemText(string textFragment)
    {
        Ui.ShouldHaveText("items-list", textFragment);
        return this;
    }

    [AllureStep("Verify items panel shows a readable error: {textFragment}")]
    public HomePage ShouldShowItemsError(string textFragment)
    {
        Ui.ShouldHaveText("items-list", textFragment);
        return this;
    }

    [AllureStep("Verify health panel shows a readable error: {textFragment}")]
    public HomePage ShouldShowHealthError(string textFragment)
    {
        Ui.ShouldHaveText("health-status", textFragment);
        return this;
    }

    [AllureStep("Verify welcome message: {message}")]
    public HomePage ShouldHaveWelcomeMessage(string message)
    {
        Ui.ShouldBeVisible("welcome-panel");
        Ui.ShouldHaveText("welcome-message", message);
        return this;
    }

    [AllureStep("Verify session panel offers logout and delete account")]
    public HomePage ShouldShowSessionActions()
    {
        Ui.ShouldBeVisible("logout-button");
        Ui.ShouldHaveText("logout-button", "Logout");
        Ui.ShouldBeVisible("delete-account-button");
        Ui.ShouldHaveText("delete-account-button", "Delete account");
        return this;
    }

    [AllureStep("Click logout button")]
    public LoginPage ClickLogoutButton()
    {
        Ui.Click("logout-button");
        return new LoginPage();
    }

    [AllureStep("Click delete account and confirm")]
    public LoginPage ClickDeleteAccountAndConfirm()
    {
        Ui.Click("delete-account-button");
        Ui.Confirm(DeleteAccountConfirm);
        return new LoginPage();
    }

    [AllureStep("Click delete account and cancel the confirm")]
    public HomePage ClickDeleteAccountAndCancel()
    {
        Ui.Click("delete-account-button");
        Ui.Dismiss(DeleteAccountConfirm);
        return this;
    }

    [AllureStep("Verify auth token remains in localStorage")]
    public HomePage ShouldKeepAuthToken()
    {
        Ui.WaitUntil(_ => Ui.Js("return localStorage.getItem(arguments[0]);", AuthTokenKey()) != null ? true : (bool?)null);
        return this;
    }
}
